src/WebJobs.Script.WebHost/Metrics/FlexConsumptionMetricsPublisher.cs

```diff
@@ -0,0 +1,301 @@
+
+            var metricsDirectoryInfo = _fileSystem.DirectoryInfo.FromDirectoryName(MetricsFilePath);
+            var files = metricsDirectoryInfo.GetFiles().OrderBy(p => p.CreationTime).ToList();
+
+            // ensure we're under the max file count
+            if (files.Count < _optionsMonitor.CurrentValue.MaxFileCount)
+            {
+                return true;
+            }
+
+            // we're at or over limit
+            // delete enough files that we have space to write a new one
+            int numToDelete = files.Count - _optionsMonitor.CurrentValue.MaxFileCount + 1;
+            var filesToDelete = files.Take(numToDelete).ToArray();
+
+            _logger.LogDebug($"Deleting {filesToDelete.Length} metrics file(s).");
+
+            foreach (var file in filesToDelete)
+            {
+                try
+                {
+                    file.Delete();
+                }
+                catch (Exception ex) when (!ex.IsFatal())
+                {
+                    // best effort
+                    _logger.LogError(ex, $"Error deleting metrics file '{file.FullName}'.");
+                }
+            }
+
+            files = metricsDirectoryInfo.GetFiles().OrderBy(p => p.CreationTime).ToList();
+
+            // return true if we have space for a new file
+            return files.Count < _optionsMonitor.CurrentValue.MaxFileCount;
+        }
+
+        private void OnStandbyOptionsChange()
+        {
+            if (!_standbyOptions.CurrentValue.InStandbyMode)
+            {
+                Start();
+            }
+        }
+
+        public void OnFunctionStarted(string functionName, string invocationId)
+        {
+            if (!_started)
+            {
+                return;
+            }
+
+            lock (_lock)
+            {
+                if (ActiveFunctionCount == 0)
+                {
+                    // we're transitioning from inactive to active
+                    _instanceActivityStopwatch = ValueStopwatch.StartNew();
+                }
+
+                ActiveFunctionCount++;
+            }
+        }
+
+        public void OnFunctionCompleted(string functionName, string invocationId)
+        {
+            if (!_started)
+            {
+                return;
+            }
+
+            lock (_lock)
+            {
+                if (ActiveFunctionCount > 0)
+                {
+                    ActiveFunctionCount--;
+                }
+
+                if (ActiveFunctionCount == 0)
+                {
+                    // we're transitioning from active to inactive accumulate the elapsed time,
+                    // applying the minimum interval
+                    var elapsedMS = _instanceActivityStopwatch.GetElapsedTime().TotalMilliseconds;
+                    var duration = Math.Max(elapsedMS, _optionsMonitor.CurrentValue.MinimumActivityIntervalMS);
+                    FunctionExecutionTimeMS += (long)duration;
```

Comment:
Question for my learning:
What does the `FunctionExecutionTimeMS` prop represents? Based on the logic above, it looks like a super set of execution time for a bunch of executions.

Let's say we have 3 function executions.

- A started at 10:00:00 and finished at 10:00:40
- B started at 10:00:00 and finished at 10:00:20
- C started at 10:00:15 and finished at 10:00:45

So the logic above will generate 45 seconds as the value of `elapsedMS`. What does that value represent? Where & how the value of `FunctionExecutionTimeMS` eventually used (after it gets published)?

Reply:
Also what happens one execution is still in progress while `OnFunctionMetricsPublishTimer` is triggered?

Reply:
Yes, as you say FunctionExecutionTimeMS is just the total time during a metering window where the host was running one or more functions. I tried to explain that in the [Metrics.ExecutionTimeMS](https://github.com/Azure/azure-functions-host/blob/08ab9671fb783d3d553b27a16253939d6d287f28/src/WebJobs.Script.WebHost/Metrics/FlexConsumptionMetricsPublisher.cs#L286) property doc.

For your second question, OnPublishMetrics resets the activity timer each interval, so tracking starts anew.